astropy/utils/data.py

```diff
@@ -268,6 +268,8 @@ def get_readable_fileobj(name_or_obj, encoding=None, cache=False,
                 name_or_obj, cache=cache, show_progress=show_progress,
                 timeout=remote_timeout, sources=sources,
                 http_headers=http_headers)
+        else:
+            name_or_obj = os.path.expanduser(name_or_obj)
```

Comment:
I'm assuming this was needed to get the `io.ascii` changes working, but this change is in widely-used core library that will impact not only other parts of astropy core but potentially other affiliated packages.

I'm not sure this change should be made here without getting wider concurrence from the community, e.g. an email on astropy-dev. Maybe nobody cares or everyone likes it.

Reply:
I had that thought about the "what's new", too. However, @svank is addressing #12778 is a number of PRs per sub-package. This one will make it into this release, but the other's won't. Personally, I think it's cleaner to announce it in what's new when it works for most of astropy, but I don't have strong feelings either way. 

Adding more tests is always good though.

Reply:
My gut feeling about paths with literal tildes is also that this is very unlikely to break things, especially since the path must start with either `~/` or `~<valid system username here>/`, since `~<string that's not a valid username>` is left unchanged.

But @taldcroft 's point that this one change could impact parts of astropy outside `io` or other packages is a good one, especially if this brings inconsistent tilde expansion to those areas. For instance, in this PR, `expanduser` had to be added in `utils/data.py` and also in `io/ascii/ui.py`, so only having the `utils/data.py` change would bring incomplete support to this module. As another example, for the tilde-path support for `io.fits` in #13131, I'd first tried adding `expanduser` to the low-level function in `fits` that calls `open()`, but that didn't expand the path for the preceding "does this file already exist?" checks in some functions, and so `expanduser` had to occur earlier in the call stack. I could imagine a similar problem for other modules/packages, where the tilde is expanded in `utils/data.py` when the file is opened, but not for any preceding, function-specific checks.

Now that I'm thinking about this, I think the right choice is to move `expanduser` out of `get_readable_fileobj` in `utils/data.py` and into the functions in `io` that call `get_readable_fileobj`.

Reply:
> Now that I'm thinking about this, I think the right choice is to move expanduser out of get_readable_fileobj in utils/data.py and into the functions in io that call get_readable_fileobj.

This seems like a good plan for the 5.1 release to contain the impact and reduce inconsistencies in the Astropy ecosystem.